kernel/src/main/scala/cats/kernel/Eq.scala

```diff
@@ -0,0 +1,138 @@
+package cats.kernel
+
+import scala.{specialized => sp}
+
+import scala.math.Equiv
+
+/**
+ * A type class used to determine equality between 2 instances of the same
+ * type. Any 2 instances `x` and `y` are equal if `eqv(x, y)` is `true`.
+ * Moreover, `eqv` should form an equivalence relation.
+ */
+trait Eq[@sp A] extends Any with Serializable { self =>
+
+  /**
+   * Returns `true` if `x` and `y` are equivalent, `false` otherwise.
+   */
+  def eqv(x: A, y: A): Boolean
+
+  /**
+   * Returns `false` if `x` and `y` are equivalent, `true` otherwise.
+   */
+  def neqv(x: A, y: A): Boolean = !eqv(x, y)
+
+  /**
+   * Constructs a new `Eq` instance for type `B` where 2 elements are
+   * equivalent iff `eqv(f(x), f(y))`.
+   */
+  def on[@sp B](f: B => A): Eq[B] =
+    new Eq[B] {
+      def eqv(x: B, y: B): Boolean = self.eqv(f(x), f(y))
+    }
+
+  /**
+   * Return an Eq that gives the result of the and of this and that
+   * note this is idempotent
+   */
+  def and(that: Eq[A]): Eq[A] =
+    new Eq[A] {
+      def eqv(x: A, y: A) = self.eqv(x, y) && that.eqv(x, y)
+    }
+  /**
+   * Return an Eq that gives the result of the or of this and that
+   * Note this is idempotent
+   */
+  def or(that: Eq[A]): Eq[A] =
+    new Eq[A] {
+      def eqv(x: A, y: A) = self.eqv(x, y) || that.eqv(x, y)
+    }
+}
+
+trait EqFunctions {
```

Comment:
Personal preference is that we leave them unsealed, but abstract classes. Making them abstract classes makes them slightly less flexible, but since the whole purpose of cats-kernel is to be stable, I think it's a trade-off worth making. Libraries higher up the stack can make their `*Functions` traits to allow for multiple inheritance, but it isn't needed here.